File: vfmd_system/src/physics_calculator.py
"""
VFMD Physics Calculators
Computes PEG, TI, and Coherence from vector fields
"""

import numpy as np
from typing import Tuple


class PhysicsCalculator:
    """Compute physical quantities from market vector fields"""
    
    @staticmethod
    def compute_peg(field, region_size=10):
        """
        Potential Energy Gradient (PEG)
        
        Measures stored energy before directional release
        PEG = ∫|∇F| dA over recent region
        
        Args:
            field: Vector field (spatial, temporal, 2)
            region_size: How many recent bars to integrate over
            
        Returns:
            peg_value: Scalar energy level
        """
        # Calculate field gradient magnitude
        grad_Fx_spatial = np.gradient(field[:, :, 0], axis=0)
        grad_Fx_temporal = np.gradient(field[:, :, 0], axis=1)
        grad_Fy_spatial = np.gradient(field[:, :, 1], axis=0)
        grad_Fy_temporal = np.gradient(field[:, :, 1], axis=1)
        
        # Total gradient magnitude at each point
        gradient_magnitude = np.sqrt(
            grad_Fx_spatial**2 + grad_Fx_temporal**2 +
            grad_Fy_spatial**2 + grad_Fy_temporal**2
        )
        
        # Integrate over recent region (right edge = most recent)
        recent_region = gradient_magnitude[:, -region_size:]
        peg = np.sum(np.abs(recent_region))
        
        # Normalize by region size for consistency
        peg = peg / region_size
        
        return peg
    
    @staticmethod
    def compute_ti(field, local_window=10):
        """
        Turbulence Index (TI)
        
        Measures chaotic instability in flow
        TI = Var(local vector angles) / Mean(directional coherence)
        
        Args:
            field: Vector field
            local_window: Recent bars to analyze
            
        Returns:
            ti_value: Turbulence level
        """
        # Extract recent region
        recent_field = field[:, -local_window:, :]
        
        # Calculate vector angles
        angles = np.arctan2(recent_field[:, :, 1], recent_field[:, :, 0])
        
        # Variance of angles (how inconsistent are directions?)
        angle_variance = np.var(angles)
        
        # Vector magnitudes
        magnitudes = np.linalg.norm(recent_field, axis=2)
        mean_magnitude = np.mean(magnitudes)
        
        # Directional coherence: how aligned are vectors?
        mean_vector = np.mean(recent_field, axis=(0, 1))
        coherence = np.linalg.norm(mean_vector) / (mean_magnitude + 1e-8)
        
        # TI = chaos / alignment
        ti = angle_variance / (coherence + 1e-8)
        
        return ti
    
    @staticmethod
    def compute_coherence(field, window=10):
        """
        Directional Coherence
        
        How strongly field points in consistent direction
        
        Args:
            field: Vector field
            window: Recent bars
            
        Returns:
            (coherence_magnitude, dominant_angle)
        """
        recent_field = field[:, -window:, :]
        
        # Average vector
        mean_vector = np.mean(recent_field, axis=(0, 1))
        
        # Coherence = magnitude of average vector
        coherence = np.linalg.norm(mean_vector)
        
        # Dominant direction
        angle = np.arctan2(mean_vector[1], mean_vector[0])
        
        # Normalize coherence to [0, 1]
        max_possible = np.max(np.linalg.norm(recent_field, axis=2))
        coherence_normalized = coherence / (max_possible + 1e-8)
        
        return coherence_normalized, angle
    
    @staticmethod
    def compute_all_metrics(field):
        """
        Convenience function: compute all physics metrics at once
        
        Returns:
            dict with keys: peg, ti, coherence, angle, divergence, curl
        """
        from field_constructor import FieldAnalyzer
        analyzer = FieldAnalyzer()
        
        peg = PhysicsCalculator.compute_peg(field)
        ti = PhysicsCalculator.compute_ti(field)
        coherence, angle = PhysicsCalculator.compute_coherence(field)
        divergence = analyzer.compute_divergence(field)
        curl = analyzer.compute_curl(field)
        
        return {
            'peg': peg,
            'ti': ti,
            'coherence': coherence,
            'angle': angle,
            'divergence_map': divergence,
            'curl_map': curl,
            'recent_divergence': divergence[:, -10:].mean(),
            'recent_curl': np.abs(curl[:, -10:]).mean()
        }


class ValidationEngine:
    """Validates physics metrics against known market behavior"""
    
    @staticmethod
    def validate_peg_on_breakouts(prices, breakout_indices, lookback=20):
        """
        Test: Does PEG spike before breakouts?
        
        Args:
            prices: Price array
            breakout_indices: List of breakout bar indices
            lookback: Bars before breakout to check
            
        Returns:
            dict with validation results
        """
        from field_constructor import SimpleFieldConstructor
        
        constructor = SimpleFieldConstructor(
            spatial_bins=50,
            temporal_window=100
        )
        
        results = []
        
        for breakout_idx in breakout_indices:
            # Need enough data before breakout
            if breakout_idx < 120:
                continue
            
            # Calculate PEG timeline leading to breakout
            peg_timeline = []
            
            for offset in range(lookback, 0, -1):
                idx = breakout_idx - offset
                price_window = prices[idx-100:idx]
                
                try:
                    field = constructor.construct_field(price_window)
                    peg = PhysicsCalculator.compute_peg(field)
                    peg_timeline.append(peg)
                except:
                    continue
            
            if len(peg_timeline) < lookback // 2:
                continue
            
            # Check if PEG was rising in the 5-15 bars before breakout
            early_peg = np.mean(peg_timeline[:5])  # Bars 20-15 before
            mid_peg = np.mean(peg_timeline[5:10])   # Bars 15-10 before
            late_peg = np.mean(peg_timeline[10:15]) # Bars 10-5 before
            
            peg_rising = (mid_peg > early_peg) and (late_peg > mid_peg)
            peg_max = np.max(peg_timeline[-15:])  # Max in last 15 bars
            peg_max_timing = lookback - np.argmax(peg_timeline[-15:])
            
            results.append({
                'breakout_idx': breakout_idx,
                'peg_rising': peg_rising,
                'peg_max': peg_max,
                'peg_max_lead_bars': peg_max_timing,
                'peg_timeline': peg_timeline
            })
        
        # Summary statistics
        n_total = len(results)
        n_rising = sum(r['peg_rising'] for r in results)
        avg_lead = np.mean([r['peg_max_lead_bars'] for r in results])
        
        print(f"\n{'='*60}")
        print("PEG VALIDATION ON BREAKOUTS")
        print(f"{'='*60}")
        print(f"Breakouts analyzed: {n_total}")
        print(f"PEG rising before breakout: {n_rising}/{n_total} ({n_rising/n_total*100:.1f}%)")
        print(f"Average PEG spike lead time: {avg_lead:.1f} bars")
        print(f"{'='*60}\n")
        
        # Success criteria: >70% of breakouts had rising PEG
        success = (n_rising / n_total) > 0.7 if n_total > 0 else False
        
        return {
            'success': success,
            'peg_prediction_rate': n_rising / n_total if n_total > 0 else 0,
            'avg_lead_bars': avg_lead,
            'details': results
        }
    
    @staticmethod
    def validate_ti_on_chop(prices, chop_zone_indices, window=50):
        """
        Test: Is TI elevated during choppy periods?
        
        Args:
            prices: Price array
            chop_zone_indices: List of chop zone center indices
            window: Bars to analyze
            
        Returns:
            dict with validation results
        """
        from field_constructor import SimpleFieldConstructor
        
        constructor = SimpleFieldConstructor(
            spatial_bins=50,
            temporal_window=100
        )
        
        chop_ti_values = []
        trend_ti_values = []
        
        for chop_idx in chop_zone_indices:
            if chop_idx < 150 or chop_idx > len(prices) - 150:
                continue
            
            # Calculate TI during chop
            chop_prices = prices[chop_idx-100:chop_idx]
            try:
                field = constructor.construct_field(chop_prices)
                ti = PhysicsCalculator.compute_ti(field)
                chop_ti_values.append(ti)
            except:
                continue
            
            # Calculate TI during trend (before chop)
            trend_prices = prices[chop_idx-250:chop_idx-150]
            try:
                field = constructor.construct_field(trend_prices)
                ti = PhysicsCalculator.compute_ti(field)
                trend_ti_values.append(ti)
            except:
                continue
        
        avg_chop_ti = np.mean(chop_ti_values) if chop_ti_values else 0
        avg_trend_ti = np.mean(trend_ti_values) if trend_ti_values else 0
        
        print(f"\n{'='*60}")
        print("TI VALIDATION ON CHOP ZONES")
        print(f"{'='*60}")
        print(f"Chop zones analyzed: {len(chop_ti_values)}")
        print(f"Average TI during chop: {avg_chop_ti:.2f}")
        print(f"Average TI during trend: {avg_trend_ti:.2f}")
        print(f"TI elevation ratio: {avg_chop_ti/avg_trend_ti:.2f}x")
        print(f"{'='*60}\n")
        
        # Success criteria: Chop TI should be >1.5x trend TI
        success = (avg_chop_ti / avg_trend_ti) > 1.5 if avg_trend_ti > 0 else False
        
        return {
            'success': success,
            'avg_chop_ti': avg_chop_ti,
            'avg_trend_ti': avg_trend_ti,
            'ti_ratio': avg_chop_ti / avg_trend_ti if avg_trend_ti > 0 else 0
        }
    
    @staticmethod
    def validate_coherence(prices, sample_size=20):
        """
        Test: Is coherence high during trends, low during ranges?
        
        Args:
            prices: Price array
            sample_size: Number of periods to test
        """
        from field_constructor import SimpleFieldConstructor
        
        constructor = SimpleFieldConstructor(
            spatial_bins=50,
            temporal_window=100
        )
        
        # Find trending periods (large net movement)
        trend_coherence = []
        range_coherence = []
        
        for i in range(200, len(prices)-100, len(prices)//sample_size):
            window = prices[i-100:i]
            
            # Calculate net movement
            net_movement = abs(window[-1] - window[0]) / window[0]
            
            try:
                field = constructor.construct_field(window)
                coherence, _ = PhysicsCalculator.compute_coherence(field)
                
                if net_movement > 0.03:  # >3% move = trend
                    trend_coherence.append(coherence)
                elif net_movement < 0.01:  # <1% move = range
                    range_coherence.append(coherence)
            except:
                continue
        
        avg_trend_coh = np.mean(trend_coherence) if trend_coherence else 0
        avg_range_coh = np.mean(range_coherence) if range_coherence else 0
        
        print(f"\n{'='*60}")
        print("COHERENCE VALIDATION")
        print(f"{'='*60}")
        print(f"Trending periods: {len(trend_coherence)}")
        print(f"Ranging periods: {len(range_coherence)}")
        print(f"Avg coherence in trends: {avg_trend_coh:.3f}")
        print(f"Avg coherence in ranges: {avg_range_coh:.3f}")
        print(f"Ratio (should be >1.5): {avg_trend_coh/avg_range_coh:.2f}x")
        print(f"{'='*60}\n")
        
        success = (avg_trend_coh / avg_range_coh) > 1.3 if avg_range_coh > 0 else False
        
        return {
            'success': success,
            'trend_coherence': avg_trend_coh,
            'range_coherence': avg_range_coh
        }


if __name__ == "__main__":
    print("Physics Calculator Module")
    print("Use ValidationEngine to test on real data")

File: server/services/vfmd/physicsCalculator.ts
/**
 * VFMD Physics Calculator - TypeScript Port
 * Computes PEG (Potential Energy Gradient), TI (Turbulence Index), and Coherence
 */

import type { VectorField, PhysicsMetrics } from './types';
import { FieldAnalyzer } from './fieldConstructor';
import TriggerCalculator from './triggerCalculator';

export class PhysicsCalculator {
  /**
   * Compute Potential Energy Gradient (PEG)
   * Delegates to centralized TriggerCalculator.computeCompressionPEG for consistency
   * @param field Vector field
   * @param regionSize How many recent bars to integrate over
   * @returns PEG value, range [0, 1]
   */
  static computePEG(field: VectorField, regionSize: number = 10): number {
    return TriggerCalculator.computeCompressionPEG(field, regionSize);
  }

  /**
   * Compute Turbulence Index (TI)
   *
   * Measures chaotic instability in flow
   * TI = Var(local vector angles) / Mean(directional coherence)
   *
   * @param field Vector field
   * @param localWindow Recent bars to analyze
   * @returns TI value (higher = more turbulent/chaotic)
   */
  static computeTurbulenceIndex(
    field: VectorField,
    localWindow: number = 10
  ): number {
    const data = field.data;
    const startT = Math.max(0, field.temporalWindow - localWindow);

    // Collect angles and magnitudes from recent region
    const angles: number[] = [];
    const magnitudes: number[] = [];

    for (let i = 0; i < field.spatialBins; i++) {
      for (let j = startT; j < field.temporalWindow; j++) {
        const [fx, fy] = data[i][j];
        const angle = Math.atan2(fy, fx);
        const magnitude = Math.sqrt(fx * fx + fy * fy);

        angles.push(angle);
        magnitudes.push(magnitude);
      }
    }

    if (angles.length === 0) return 0;

    // Variance of angles (how inconsistent are directions?)
    const meanAngle =
      angles.reduce((a, b) => a + b, 0) / angles.length;
    const angleVariance =
      angles.reduce((sum, a) => sum + Math.pow(a - meanAngle, 2), 0) /
      angles.length;

    // Mean magnitude
    const meanMagnitude =
      magnitudes.reduce((a, b) => a + b, 0) / magnitudes.length;

    // Guard: pure flat market (zero magnitude) → zero turbulence
    if (meanMagnitude < 1e-8) return 0;

    // Directional coherence: how aligned are vectors?
    let meanVectorX = 0;
    let meanVectorY = 0;
    for (let i = 0; i < field.spatialBins; i++) {
      for (let j = startT; j < field.temporalWindow; j++) {
        const [fx, fy] = data[i][j];
        meanVectorX += fx;
        meanVectorY += fy;
      }
    }

    const count = field.spatialBins * (field.temporalWindow - startT);
    meanVectorX /= count;
    meanVectorY /= count;

    const coherence =
      Math.sqrt(meanVectorX * meanVectorX + meanVectorY * meanVectorY) /
      (meanMagnitude + 1e-8);

    // TI = chaos / alignment
    const ti = angleVariance / (coherence + 1e-8);

    // Cap TI at 10.0 — extreme TI values (587+) are measurement artifacts
    // preventing incorrect turbulent-regime classification
    return Math.min(10.0, Math.max(0, ti));
  }

  /**
   * Compute Directional Coherence
   *
   * How strongly field points in consistent direction
   *
   * @param field Vector field
   * @param window Recent bars
   * @returns [coherenceScore, dominantAngle]
   */
  static computeCoherence(
    field: VectorField,
    window: number = 10
  ): [number, number] {
    const data = field.data;
    const startT = Math.max(0, field.temporalWindow - window);

    // Calculate mean vector
    let meanX = 0;
    let meanY = 0;
    let maxMagnitude = 0;

    for (let i = 0; i < field.spatialBins; i++) {
      for (let j = startT; j < field.temporalWindow; j++) {
        const [fx, fy] = data[i][j];
        meanX += fx;
        meanY += fy;

        const mag = Math.sqrt(fx * fx + fy * fy);
        maxMagnitude = Math.max(maxMagnitude, mag);
      }
    }

    const count = field.spatialBins * (field.temporalWindow - startT);
    meanX /= count;
    meanY /= count;

    const coherence = Math.sqrt(meanX * meanX + meanY * meanY);
    const coherenceNormalized = coherence / (maxMagnitude + 1e-8);

    const angle = Math.atan2(meanY, meanX);

    return [
      Math.min(1, Math.max(0, coherenceNormalized)),
      angle
    ];
  }

  /**
   * Compute all physics metrics at once
   */
  static computeAllMetrics(field: VectorField): PhysicsMetrics {
    const peg = this.computePEG(field);
    const ti = this.computeTurbulenceIndex(field);
    const [coherence, angle] = this.computeCoherence(field);

    const divergence = FieldAnalyzer.computeDivergence(field);
    const curl = FieldAnalyzer.computeCurl(field);

    // Recent statistics (last 10 bars)
    const startT = Math.max(0, field.temporalWindow - 10);

    let recentDivSum = 0;
    let recentCurlSum = 0;
    let count = 0;

    for (let i = 0; i < field.spatialBins; i++) {
      for (let j = startT; j < field.temporalWindow; j++) {
        recentDivSum += divergence[i][j];
        recentCurlSum += Math.abs(curl[i][j]);
        count++;
      }
    }

    const recentDiv = recentDivSum / (count || 1);
    const recentCurl = recentCurlSum / (count || 1);

    // Overall gradient magnitude
    const gradMag = FieldAnalyzer.computeGradientMagnitude(field);
    let maxGrad = 0;
    for (const row of gradMag) {
      for (const val of row) {
        maxGrad = Math.max(maxGrad, val);
      }
    }

    return {
      peg: Math.max(0, peg),
      turbulenceIndex: Math.max(0, ti),
      coherenceScore: coherence,
      dominantAngle: angle,
      divergenceScore: recentDiv,
      recentDivergence: recentDiv,
      curlScore: recentCurl,
      recentCurl,
      gradientMagnitude: maxGrad
    };
  }

  /**
   * Compute PEG derivatives for three-tier signal quality gating
   * 
   * Returns physics-based signal quality tier:
   * EXPLOSIVE:  PEG > 0.70× threshold AND ΔPEG > 0.03 AND Δ²PEG > 0
   *            → True breakout onset (pressure present, building, accelerating)
   *            → Position multiplier: 1.3x (highest conviction)
   * 
   * BUILDING:   PEG > 0.85× threshold AND ΔPEG > 0.03
   *            → Good entries (compression above floor, still accumulating)
   *            → Position multiplier: 1.1x (high conviction)
   * 
   * BASE:       PEG > threshold AND isBuilding = false
   *            → Current system default (still valid but lower conviction)
   *            → Position multiplier: 0.6x (reduced for spent energy)
   * 
   * @param pegHistory Rolling buffer of recent PEG values (needs ≥3 points)
   * @param pegThreshold Current regime's PEG threshold for signal gating
   * @returns { tier, multiplier, deltaPeg, delta2Peg, isBuilding, isExplosive }
   */
  static computePEGDerivatives(
    pegHistory: number[],
    pegThreshold: number = 0.20
  ): {
    tier: 'EXPLOSIVE' | 'BUILDING' | 'BASE';
    multiplier: number;
    deltaPeg: number;
    delta2Peg: number;
    pegMomentum: number;
    isBuilding: boolean;
    isExplosive: boolean;
  } {
    // Need at least 3 points for first and second derivatives
    if (pegHistory.length < 3) {
      return {
        tier: 'BASE',
        multiplier: 0.6,
        deltaPeg: 0,
        delta2Peg: 0,
        pegMomentum: 0,
        isBuilding: false,
        isExplosive: false
      };
    }

    // Use last 5 points for noise tolerance, EMA-smooth if available
    const recent = pegHistory.slice(-5);
    
    // First derivative: ΔPEG (jerk — is acceleration changing?)
    // Use last 3 points for stability
    const peg2 = recent[recent.length - 2];
    const peg1 = recent[recent.length - 1];
    const peg0 = recent[recent.length - 3];
    
    const deltaPeg = peg1 - peg0;  // First derivative (momentum)
    
    // Second derivative: Δ²PEG (snap — is jerk accelerating?)
    // Needs 3 consecutive deltas
    const delta2Peg = (peg1 - peg2) - (peg2 - peg0);

    // Combined momentum scalar: sign(ΔPEG) × √|ΔPEG|
    const pegMomentum = Math.sign(deltaPeg) * Math.sqrt(Math.abs(deltaPeg));

    // Current PEG (last value)
    const currentPeg = peg1;
    const pegNorm = currentPeg / Math.max(pegThreshold, 0.01); // Normalize to threshold

    // Gate thresholds
    const isExplosive = 
      pegNorm > 0.70 &&          // PEG > 70% of threshold
      deltaPeg > 0.03 &&         // Strong upward momentum
      delta2Peg > 0;             // Acceleration is positive

    const isBuilding = 
      !isExplosive &&
      pegNorm > 0.85 &&          // PEG > 85% of threshold
      deltaPeg > 0.03;           // Building momentum

    const tier = isExplosive ? 'EXPLOSIVE' : isBuilding ? 'BUILDING' : 'BASE';
    const multiplier = isExplosive ? 1.3 : isBuilding ? 1.1 : 0.6;

    return {
      tier,
      multiplier,
      deltaPeg,
      delta2Peg,
      pegMomentum,
      isBuilding,
      isExplosive
    };
  }
}
